### PR TITLE
Use `std::filesystem::path` in end executables

### PR DESCRIPTION
**Pull request by siddhantk232**
# Motivation

Inspired from https://github.com/NixOS/nix/pull/11218. This PR tries to use `std::filesystem::path` in end executables. This reduces the amount of change in https://github.com/siddhantk232/nix/tree/more-std-filepath (by a very small amount).

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Comment by edolstra** [comment]
Hm, what is an end executable?

File: src/nix-channel/nix-channel.cc [hunk 1]
```diff
@@ -165,7 +165,7 @@ static int main_nix_channel(int argc, char ** argv)
     {
         // Figure out the name of the `.nix-channels' file to use
         auto home = getHome();
-        channelsList = settings.useXDGBaseDirectories ? createNixStateDir() + "/channels" : home + "/.nix-channels";
+        channelsList = settings.useXDGBaseDirectories ? createNixStateDir() + OS_STR("/channels") : home + OS_STR("/.nix-channels");
```
Comment (Ericson2314) [inline src/nix-channel/nix-channel.cc:168]: Was this change necessary for the windows build? Otherwise I think we actually *don't* want this yet because it will break the windows build.

Comment (siddhantk232) [reply src/nix-channel/nix-channel.cc]: I've removed `OS_STR` in https://github.com/NixOS/nix/pull/11281/commits/9ccad9708eac3593defab8a8f74d3acf1bb60870

**Comment by siddhantk232** [comment]
> Hm, what is an end executable?

I used it to mean `src/nix/*.cc` files.

**Comment by Ericson2314** [comment]
I didn't see what was causing the timeout in the logs, so I am just restarting
